### ISSUE TITLE
Problems with the new 7.5 / B.1.2 Echo procedure and GCM

### ISSUE THREAD

**Issue by emanjon** [core-wg/oscoap#251]
Hi,

I do not think that the new suggested Echo procedure specified in 7.5 and B.1.2 is secure with GCM. The procedure intentionally reuses a nonce. As the Plaintext and/or AAD changes, I think the nonce reuse will allow an attacker to calculate the GCM authentication key H, therefore compromising integrity for all following messages the server receives. I don't think we want a procedure that is dependant on the AEAD algorithm.

Are there any ways to fix this or should we revert to the previous procedure where the server first need to make sure that it uses a fresh nonce and then send Echo?

What do you think @jimsch ?

In Section 7.5, bullet 2

`The endpoints can update the server’s Replay Window by using the Echo option [I-D.ietf-core-echo-request-tag], see Appendix B.1.2 for an example. This method only updates the Replay Window and does not update the server’s Sender Sequence Number, so it MUST be complemented by a method for updating the Sender Sequence Number (e.g. by using Appendix B.1.1) if the application allows for the server to encode the AEAD nonce using its Sender Sequence Number in responses to regular requests, or if the server supports Observe. This method MUST NOT be used unless the client can guarantee that its own Sender Sequence Numbers are not reused (e.g. by using Appendix B.1.1).
`
In Appendix B.1.2

```
In case of loss of security context on the server, to prevent accepting replay of previously received requests, the server may perform the following procedure after boot:

For each stored security context, the first time after boot the server receives an OSCORE request, the server responds with a 4.01 (Unauthorized) message, containing only the Echo option [I-D.ietf-core-echo-request-tag] and no diagnostic payload, in order to get a request with verifiable freshness. The server MUST use a fixed Sender Sequence Number = 2^40 - 1 when generating the AEAD nonce and MUST include it as Partial IV in the response. The server MUST set the maximum Sender Sequence Number to 2^40 - 2, and MUST NOT use 2^40 - 1 to generate the AEAD nonce for any other response.
If the server using the Echo option can verify a second request as fresh, then the Partial IV of the second request is set as the lower limit of the replay window.

This method allows the re-use of the AEAD nonce generated from the server’s Sender Sequence Number = 2^40 - 1: this is acceptable as the only changing information in the 4.01 (Unauthorized) error message is the value of the Echo option.
```

**Comment by jimsch** [comment]
Yes you are correct.  I was only thinking of the use of CCM which does not have this problem.

**Comment by gselander** [comment]
I merged PR #252 - reverting to the processing in -15.

**Comment by gselander** [comment]
I close this issue now, if there are still comments we can reopen it.
